### PR TITLE
feat:add iqcs features

### PR DESCRIPTION
**Pull request by lss0208**
need to deeply merge further, e.g. oracle, variational framework, etc.

### CODE REVIEW

File: src/qfvm/statevector.hpp [hunk 1]
```diff
@@ -1399,3 +1407,39 @@ void StateVector<real_t>::apply_reset(vector<pos_t> const& qbits) {
   const auto meas = sample_measure_probs(qbits);
   update(qbits, 0, meas.first, meas.second);
 }
+
+
+ //sample results
+template <class real_t>
+std::unordered_map<std::string, int> StateVector<real_t>::measure_samples(vector<std::pair<uint, uint>> meas, int shots){
```
Comment (Zhaoyilunnn) [inline src/qfvm/statevector.hpp:1414]: better to use `const std::vector<std::pair<int, int>> &` for non-modifiable complex data types

File: src/qfvm/statevector.hpp [hunk 2]
```diff
@@ -1041,6 +1047,8 @@ vector<double> StateVector<real_t>::probabilities() const {
   return probs;
 }
 
+
+
 vector<std::complex<double>> convert(const vector<std::complex<double>>& v) {
```
Comment (Zhaoyilunnn) [inline src/qfvm/statevector.hpp:1052]: Minor: I just find that there's inconsistency style when using const references, sometimes `xx const &` and sometimes `const xx &`, although thery are equivalent

**Review by Zhaoyilunnn — COMMENTED** [review]
Overall this PR looks good to me. 
Left some comments
- Minor issue can be neglected.
- For gradients, should we implement classes from the long term perspective? Also we already have a naive parameter shift under https://github.com/ScQ-Cloud/pyquafu/tree/f0345be8bd949a75e92ab2cd11fe82c3ee0adc50/quafu/algorithms/gradients. I will try to use your implementation and modify `vjp.py`

File: quafu/algorithms/gradient.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            circ = QuantumCircuit(max(op.pos)+1)
+            deriv_mat = -0.5j * XMatrix @ op._get_targ_matrix()
+            circ << QuantumGate("dRX", op.targs, [], deriv_mat)
+            cdim = 1 << (len(op.ctrls))
+            proj_mat = np.zeros((cdim, cdim))
+            proj_mat[cdim-1, cdim-1] = 1.
+            circ << QuantumGate("projCtrl", op.ctrls, [], proj_mat)
+            return circ.wrap()
+        
+        elif op._targ_name == "RY":
```
[the remark this answers is not in the document]
Comment (lss0208) [reply quafu/algorithms/gradient.py:72]: this gate derivative function may be implemented as a method of QuantumGate class

File: quafu/elements/instruction.py [hunk 4]
```diff
@@ -35,8 +34,8 @@ class Instruction(ABC):
 
     def __init__(
         self,
-        pos: PosType,
-        paras: Optional[Union[ParameterType, List[ParameterType]]] = None,
+        pos: List[int],
```
Comment (chensgit169) [inline quafu/elements/instruction.py:37]: It might be better to allow input 'pos' as int for the convenience of users.

**Review by chensgit169 — APPROVED** [review]
``elements`` related approved

**Review by Zhaoyilunnn — APPROVED** [review]

**Pull request merged**